jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+        filterstr (str): Filter string. will match against against Process.name(), Process.exe() and Process.cmdline()
+        n_instances (int, optional): Number of needed instances. Defaults to 1.
+        retry (int, optional): Number of retries to execute the command and check. Defaults to 1.
+        timout (int, optional): how long the function should wait for process to finish (first or parnet process started with the command)
+        delay (int, optional): how long the function should delay the checking process after the process finished or timeout exceeded (in case the first process started another one).
 
     Raises:
-        j.exceptions.RuntimeError: will be raised if we didn't reach number of required instances
+        j.exceptions.Runtime: will be raised if we didn't reach number of required instances.
     """
-    found = get_filtered_pids(filterstr)
     for i in range(retry):
-        if len(found) == nrinstances:
-            return
-        # print "START:%s"%cmd
-        execute(cmd)
-        time.sleep(1)
-    found = get_filtered_pids(filterstr)
-    if len(found) != nrinstances:
-        raise j.exceptions.RuntimeError(
-            "could not start %s, found %s nr of instances. Needed %s." % (cmd, len(found), nrinstances)
-        )
-
+        if isinstance(cmd, str):
+            args = shlex.split(cmd)
+        proc = psutil.Popen(args, close_fds=True, stdout=subprocess.PIPE, stdin=subprocess.PIPE, stderr=subprocess.PIPE)
+        j.logger.debug(f"attempt no: {i}")
+        j.logger.debug(f"executed cmd: {cmd}")
+        j.logger.debug(f"proc.cmdline: {proc.cmdline()}")
+        j.logger.debug(f"direct process id spawn by this command : {proc.pid}")
+        j.logger.debug(f"status: {proc.status()}")
+        j.logger.debug(f"name: {proc.name()}")
+        j.logger.debug(f"exe: {proc.exe()}")
+        try:
+            rc = proc.wait(timeout)  # makesure the process is stable
+            if rc == 0:  # executing the command succeeded but exited immediately!
+                j.logger.debug("executing the start command succeeded but exited immediately")
+            else:
+                j.logger.debug("the start command exited with error")  # the process exited with error
+                output, error_output = proc.communicate()
+                j.logger.debug(f"err: {error_output}")
+            j.logger.debug(f"return code is: {rc}")
+        except psutil.TimeoutExpired:
+            j.logger.debug("the start command still running after 2 sec")  # still running
+        # wait extra delay to allow any subprocess spawned from the process we just started to finish
+        # this may not needed in many cases
+        time.sleep(delay)
+        # TODO check based on command
+        if check_running(filterstr, min=n_instances):
+            j.logger.debug(f"found at least {n_instances} instances using the filter string: {filterstr}")
+            return True  # XXX should remove? None
```

Comment:
Should be removed

Reply:
done

> Should be removed

done after search the code in js-ng, js-sdk and js-grid.